src/Plugin/formatter/Formatter.php

```diff
@@ -162,26 +207,84 @@ protected function createCacheController($data) {
   }
 
   /**
-   * Remove the unnecessary fields from the response.
+   * Gets a cache fragments based on the data to be rendered.
    *
-   * @param string[] $limit_fields
-   *   The list of fields.
-   * @param array $output
-   *   The prepared output.
+   * @param mixed $data
+   *   The data to be rendered.
    *
-   * @return array
-   *   The filtered contents.
+   * @return ArrayCollection
+   *   The cache controller.
    */
-  public function limitFields(array $limit_fields, array $output) {
-    if (!$limit_fields) {
+  protected static function cacheFragments($data) {
+    $context = $data->getContext();
+    if (!$cache_fragments = $context['cache_fragments']) {
+      return NULL;
+    }
+    return $cache_fragments;
+  }
+
+  /**
+   * Returns only the allowed fields by filtering out the other ones.
+   *
+   * @param mixed $output
+   *   The data structure to filter.
+   * @param bool|string[] $allowed_fields
+   *   FALSE to allow all fields. An array of allowed values otherwise.
+   *
+   * @return mixed
+   *   The filtered output.
+   */
+  protected function limitFields($output, $allowed_fields = NULL) {
+    if (!isset($allowed_fields)) {
+      $request = ($resource = $this->getResource()) ? $resource->getRequest() : restful()->getRequest();
+      $input = $request->getParsedInput();
+      // Set the field limits to false if there are no limits.
+      $allowed_fields = empty($input['fields']) ? FALSE : explode(',', $input['fields']);
+    }
+    if (!is_array($output)) {
+      // $output is a simple value.
       return $output;
     }
-    foreach (array_keys($output) as $field_name) {
-      if (!in_array($field_name, $limit_fields)) {
-        unset($output[$field_name]);
+    $result = array();
+    if (ResourceFieldBase::isArrayNumeric($output)) {
+      foreach ($output as $item) {
+        $result[] = $this->limitFields($item, $allowed_fields);
       }
+      return $result;
     }
-    return $output;
+    foreach ($output as $field_name => $field_contents) {
+      if ($allowed_fields !== FALSE && !in_array($field_name, $allowed_fields)) {
+        continue;
+      }
+      $result[$field_name] = $this->limitFields($field_contents, $this->unprefixInputOptions($allowed_fields, $field_name));
+    }
+    return $result;
+  }
+
+  /**
+   * Given a prefix, return the allowed fields that apply removing the prefix.
+   *
+   * @param bool|string[] $allowed_fields
+   *   The list of allowed fields in dot notation.
+   * @param string $prefix
+   *   The prefix used to select the fields and to remove from the front.
+   *
+   * @return bool|string[]
+   *   The new allowed fields for the nested sub-request.
+   */
+  protected static function unprefixInputOptions($allowed_fields, $prefix) {
```

Comment:
Given some query param like `a,b,b.c,b.d,e,f,f.g,f.h,f.i`, this is used so we can pass only the request options `b.c,b.d` when making a sub-request to the resource for the `b` relationship.